packages/fhir-router/src/graphql/graphql.ts

```diff
@@ -442,25 +444,76 @@ const DEFAULT_MAX_DEPTH = 12;
  */
 const MaxDepthRule =
   (maxDepth: number = DEFAULT_MAX_DEPTH) =>
-  (context: ValidationContext): ASTVisitor => ({
-    Field(
-      /** The current node being visiting. */
-      node: any,
-      /** The index or key to this node from the parent node or Array. */
-      _key: string | number | undefined,
-      /** The parent immediately above this node, which may be an Array. */
-      _parent: ASTNode | readonly ASTNode[] | undefined,
-      /** The key path to get to this node from the root node. */
-      path: readonly (string | number)[]
-    ): any {
-      const depth = getDepth(path);
-      if (depth > maxDepth) {
-        const fieldName = node.name.value;
-        context.reportError(
-          new GraphQLError(`Field "${fieldName}" exceeds max depth (depth=${depth}, max=${maxDepth})`, {
-            nodes: node,
-          })
-        );
+  (context: ValidationContext): ASTVisitor =>
+    new MaxDepthVisitor(context, maxDepth);
+
+type DepthRecord = { depth: number; node?: FieldNode };
+
+class MaxDepthVisitor {
+  private context: ValidationContext;
+  private maxDepth: number;
+  private fragmentDepths: Record<string, DepthRecord>;
+
+  constructor(context: ValidationContext, maxDepth: number) {
+    this.context = context;
+    this.fragmentDepths = Object.create(null);
+    this.maxDepth = maxDepth;
+  }
+
+  OperationDefinition(node: OperationDefinitionNode): void {
+    const result = this.getDepth(...node.selectionSet.selections);
+    if (result.depth > this.maxDepth) {
+      this.context.reportError(
+        new GraphQLError(
+          `Field "${result.node?.name.value}" exceeds max depth (depth=${result.depth}, max=${this.maxDepth})`,
+          {
+            nodes: result.node,
+          }
+        )
+      );
+    }
+  }
+
+  /**
+   * Returns the depth of the GraphQL node in a query.
+   * We use field depth as the representation of depth: the number of concrete fields (not counting fragment expansions)
+   * @param nodes - The AST nodes.
+   * @returns The maximum "depth" of the nodes.
+   */
+  private getDepth(...nodes: ASTNode[]): DepthRecord {
```

Comment:
I think this method makes sense.  Thanks for digging into the graphql AST to figure this out.  Oof.

Now that we're breaking the seal on AST parsing, I'm wondering if we should add in some intelligence here for the specifics of FHIR GraphQL.

My interpretation of this function is that it's treating all "fields" the same (same as before).

Maybe we should keep that as-is for now.

I assume that in the near future, we would want to add in checks for `isCapital(node.name[0])` or `node.name.endsWith('List')` to detect "resolve by ID", "resolve by search", "resolve by connection API", etc.

Or, even better, if it's possible to get the resolver from the AST.  I don't know if that's possible though.

Is that correct?  Too much for this PR?  I'd be fine either way, just trying to mind meld.

Reply:
That's basically how it's handled in the [query complexity calculation](https://github.com/medplum/medplum/pull/4924/files#diff-c401c28cb20cf21ea2918c9ba3d7bf86cbabb0152b53aa0e8be327e7ad80b56fR518-R526) — I'm inclined to have a simple and straightforward max depth that just counts fields (since even backbone elements increase branching factor and can contribute to query complexity), but we can also raise the limit to compensate?